portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,25 +14,46 @@
 
 package com.google.sps.servlets;
 
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.gson.Gson;
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-import java.util.HashMap;
-// import java.util.ArrayList;
-import com.google.gson.Gson;
-
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that adds/retrieves comments stored in a Datastore */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-  HashMap<String, String> messages = new HashMap<String, String>();
-  // ArrayList<String> messages = new ArrayList<String>();
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  private static final int DEFAULT_MAX_NUM_COMMENTS = 7;
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     response.setContentType("application/json;");
+
+    Map<String, String> messages = new LinkedHashMap<>();
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    PreparedQuery results = datastore.prepare(query);
+
+    int maxNumComments = (int)request.getSession().getAttribute("maxNumComments");
```

Comment:
Just another note here - that the session won't have this attribute on your first page load.

no need to update now, but something to ponder for future work.